### PR TITLE
community: Add Naver chat model & embeddings

### PR DESCRIPTION
**Pull request by hyper-clova**
Reopened as a personal repo outside the organization.

## Description
- Naver HyperCLOVA X community package 
  - Add chat model & embeddings
  - Add unit test & integration test
  - Add chat model & embeddings docs
- I changed partner package(https://github.com/langchain-ai/langchain/pull/24252) to community package on this PR
- Could this embeddings(https://github.com/langchain-ai/langchain/pull/21890) be deprecated? We are trying to replace it with embedding model(**ClovaXEmbeddings**) in this PR.

Twitter handle: None. (if needed, contact with joonha.jeon@navercorp.com)

---
you can check our previous discussion below:

> one question on namespaces - would it make sense to have these in .clova namespaces instead of .naver?

I would like to keep it as is, unless it is essential to unify the package name.
(ClovaX is a branding for the model, and I plan to add other models and components. They need to be managed as separate classes.)

> also, could you clarify the difference between ClovaEmbeddings and ClovaXEmbeddings?

There are 3 models that are being serviced by embedding, and all are supported in the current PR. In addition, all the functionality of CLOVA Studio that serves actual models, such as distinguishing between test apps and service apps, is supported. The existing PR does not support this content because it is hard-coded.

### CODE REVIEW

**Comment by hyper-clova** [comment]
Hello @efriis, 

We've been preparing for this community package and waiting for review (following up our prior communication on early August via email), and I think we are all ready to go!

Could you check this PR?

**Comment by efriis** [comment]
Thanks for the ping! Could you get CI passing?

**Comment by hyper-clova** [comment]
Thanks @efriis , seems we referred to some outdated docs, got failed linting test. 
Now fixed!

**Comment by hyper-clova** [comment]
Hello @efriis , 
we've revised some snippets related to a preparation for pydantic version that caused CI test fails.
Now (hopefully) all fixed!

**Comment by hyper-clova** [comment]
Hi @efriis, hope this doesn't bother you much as things getting you busier preparing for v0.3 these days.

I was wondering if you've had a chance to look at this PR. Could you please check if this is good to merge? (it seems the master branch changes fast and not sure if test should be done for each update..)

I'm happy to make any necessary changes. Thanks for your time!

**Comment by efriis** [comment]
Howdy! Needs support per our [review process](https://python.langchain.com/v0.2/docs/contributing/review_process/)

**Comment by hyper-clova** [comment]
@efriis, is ([review process](https://python.langchain.com/v0.2/docs/contributing/review_process/)) newly made public or changed? 

I think our PR has been submitted prior to the guidelines were implemented and given this timing, I wonder if it's appropriate to apply these new standards retroactively to PRs that were already previously registered. 

Also I'm not sure where should we get upvotes from the community - PR itself or relevant issue/discussion?

File: libs/community/langchain_community/chat_models/naver.py [hunk 1]
```diff
@@ -0,0 +1,565 @@
...
+    BaseMessageChunk,
+    ChatMessage,
+    ChatMessageChunk,
+    HumanMessage,
+    HumanMessageChunk,
+    SystemMessage,
+    SystemMessageChunk,
+)
+from langchain_core.outputs import ChatGeneration, ChatGenerationChunk, ChatResult
+from langchain_core.pydantic_v1 import Field, SecretStr
```
Comment (efriis) [inline libs/community/langchain_community/chat_models/naver.py:37]: needs to be pydantic 2!

File: libs/community/langchain_community/chat_models/naver.py [hunk 2]
```diff
@@ -0,0 +1,565 @@
...
+        app_type = "serviceapp" if self.service_app else "testapp"
+
+        if self.task_id:
+            return (
+                f"{self.base_url}/{app_type}/v1/tasks/{self.task_id}/chat-completions"
+            )
+        else:
+            return f"{self.base_url}/{app_type}/v1/chat-completions/{self.model_name}"
+
+    def __init__(self, **kwargs: Any) -> None:
```
Comment (efriis) [inline libs/community/langchain_community/chat_models/naver.py:273]: let's make these pydantic validated fields instead - much cleaner code! e.g. for numeric fields: https://docs.pydantic.dev/latest/concepts/fields/#numeric-constraints

File: libs/community/langchain_community/chat_models/naver.py [hunk 3]
```diff
@@ -0,0 +1,565 @@
...
+        if "async_client" in kwargs:
+            self.async_client = kwargs.get("async_client")
+        else:
+            self.async_client = httpx.AsyncClient(
+                base_url=self.base_url,
+                headers=self.default_headers(kwargs),
+                timeout=self.timeout,
+            )
+
+    @staticmethod
```
Comment (efriis) [inline libs/community/langchain_community/chat_models/naver.py:355]: could you explain this? I don't think this is a static method if it's relying on instance properties like the api key

File: libs/community/langchain_community/embeddings/clova.py [hunk 4]
```diff
@@ -3,11 +3,17 @@
...
 import requests
+from langchain_core._api.deprecation import deprecated
 from langchain_core.embeddings import Embeddings
 from langchain_core.utils import convert_to_secret_str, get_from_dict_or_env
 from pydantic import BaseModel, ConfigDict, SecretStr, model_validator
 
 
+@deprecated(
+    since="0.2.12",
+    removal="0.3.0",
```
Comment (efriis) [inline libs/community/langchain_community/embeddings/clova.py:14]: 1.0.0

File: libs/community/langchain_community/embeddings/naver.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+import httpx
+from langchain_core.embeddings import Embeddings
+from langchain_core.pydantic_v1 import (
+    BaseModel,
+    Field,
+    SecretStr,
+)
+from langchain_core.utils import convert_to_secret_str, get_from_dict_or_env
+
+DEFAULT_BASE_URL = "https://clovastudio.apigw.ntruss.com"
```
Comment (efriis) [inline libs/community/langchain_community/embeddings/naver.py:13]: make it private _DEFAULT_BASE_URL

**Review by efriis — COMMENTED** [review]
guessing there will be a sizable number of CI failures on this - feel free to click in and see the issues! To run linting+testing yourself locally you can

```
cd libs/community
poetry install --with lint,typing --sync
make lint
```

and 
```
cd libs/community
poetry install --with lint,typing,test
make test
```

File: libs/community/langchain_community/chat_models/naver.py (hunk 1 above)
Comment (hyper-clova) [reply libs/community/langchain_community/chat_models/naver.py:37]: AS-IS
```py
from langchain_core.pydantic_v1 import Field, SecretStr
```
TO-BE
```py
from pydantic import Field, SecretStr
```

is this code right?

File: libs/community/langchain_community/embeddings/clova.py (hunk 4 above)
Comment (hyper-clova) [reply libs/community/langchain_community/embeddings/clova.py:14]: Should I change `removal` to 1.0.0?

Comment (efriis) [reply libs/community/langchain_community/embeddings/clova.py:14]: yep!

**Comment by efriis** [comment]
also pinged you on slack for an api key for testing!

File: libs/community/langchain_community/chat_models/naver.py (hunk 1 above)
Comment (efriis) [reply libs/community/langchain_community/chat_models/naver.py:37]: Yes! Will likely require some changes in the code as well

**Comment by efriis** [comment]
actually scratch the api key - let's address the above comments and get CI passing!

File: libs/community/langchain_community/chat_models/naver.py (hunk 2 above)
Comment (hyper-clova) [reply libs/community/langchain_community/chat_models/naver.py]: ok I'll change to pydantic validated fields on tomorrow(9/24).

File: libs/community/langchain_community/chat_models/naver.py (hunk 3 above)
Comment (hyper-clova) [reply libs/community/langchain_community/chat_models/naver.py:361]: OK, I'll change it to a member method instead of a static method.

**Comment by efriis** [comment]
You responded to me on Friday! Is this Joonha?

File: libs/community/langchain_community/chat_models/naver.py (hunk 2 above)
Comment (hyper-clova) [reply libs/community/langchain_community/chat_models/naver.py]: Applied.

File: libs/community/langchain_community/chat_models/naver.py (hunk 3 above)
Comment (hyper-clova) [reply libs/community/langchain_community/chat_models/naver.py]: Applied.

**Comment by hyper-clova** [comment]
> You responded to me on Friday! Is this Joonha?

Sorry for any confusion. This is our team account and there was a little communication gap between us. Checked Joonha (PM) has sent you the api key :)

**Comment by hyper-clova** [comment]
Hi @efriis , I think we missed the release v0.3.1 for langchain-community. Do you plan this to be included in v0.3.2?

+ Just noticed an issue (#26695 ) that tries to add on existing embedding component (ClovaEmbeddings) due to high demand of using langchain with our APIs, however, the existing one should be deprecated due to the reason we mentioned before.

**Comment by hyper-clova** [comment]
Hi @efriis , I have fixed some issue on your commands. Please check fixed commit.

**Comment by hyper-clova** [comment]
Hi @efriis , are there any additional tasks or changes needed before this PR can be merged? We're eager to contribute and appreciate your time reviewing this :)

File: docs/docs/integrations/chat/naver.ipynb [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "e49f1e0d",
+   "metadata": {},
+   "source": [
+    "# ChatClovaX\n",
+    "\n",
+    "This notebook provides a quick overview for getting started with Naver’s HyperCLOVA X [chat models](https://python.langchain.com/v0.2/docs/concepts/#chat-models) via CLOVA Studio. For detailed documentation of all ChatClovaX features and configurations head to the [API reference](https://api.python.langchain.com/en/latest/community/chat_models/langchain_community.chat_models.naver.ChatClovaX.html).\n",
```
Comment (efriis) [inline docs/docs/integrations/chat/naver.ipynb:20]: please link to recent docs at /docs, don't link to /v0.2 docs

Comment (efriis) [reply docs/docs/integrations/chat/naver.ipynb:20]: also api reference should start with python.langchain.com/api_reference, not api.python

File: docs/docs/integrations/chat/naver.ipynb [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+    "\n",
+    "CLOVA Studio has several chat models. You can find information about latest models and their costs, context windows, and supported input types in the CLOVA Studio API Guide [documentation](https://api.ncloud-docs.com/docs/clovastudio-chatcompletions).\n",
+    "\n",
+    "## Overview\n",
+    "### Integration details\n",
+    "\n",
+    "| Class | Package | Local | Serializable | [JS support](https://js.langchain.com/v0.2/docs/integrations/chat/naver) | Package downloads | Package latest |\n",
```
Comment (efriis) [inline docs/docs/integrations/chat/naver.ipynb:27]: this linked js integration doesn't exist

File: docs/docs/integrations/chat/naver.ipynb [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+    "| :--- | :--- |:-----:| :---: |:------------------------------------------------------------------------:| :---: | :---: |\n",
+    "| [ChatClovaX](https://api.python.langchain.com/en/latest/chat_models/langchain_community.chat_models.naver.ChatClovaX.html) | [langchain-community](https://api.python.langchain.com/en/latest/community_api_reference.html) |   ❌   | beta/❌ |                                    ❌                                     | ![PyPI - Downloads](https://img.shields.io/pypi/dm/langchain-naver?style=flat-square&label=%20) | ![PyPI - Version](https://img.shields.io/pypi/v/langchain-naver?style=flat-square&label=%20) |\n",
```
Comment (efriis) [inline docs/docs/integrations/chat/naver.ipynb:29]: you can delete this whole table. there's a lot of broken links in it.

**Review by efriis — COMMENTED** [review]
only reviewed first part of the first file (chat/naver.ipynb) - could you go through the whole PR and give it a self review to confirm links work, code works, that kind of thing?

at the moment it seems to mostly match a template that is out of date and also not accurate (e.g. it links to a langchainjs integration that doesn't exist, and documents the community integration like it's a partner package).

File: docs/docs/integrations/chat/naver.ipynb (hunk 6 above)
Comment (hyper-clova) [reply docs/docs/integrations/chat/naver.ipynb:20]: Thanks for the comments including below ones. Now self reviewing docs and codes finding what you mentioned are the most part to be modified. 

Will fix them soon, but I have a quick question to be sure as you mentioned "a template that is out of date": 
when writing docs, now I'm relying on those in [langchain_cli.integration_template](https://github.com/langchain-ai/langchain/tree/master/libs/cli/langchain_cli/integration_template), and some of the most recent updates in langchain_community like [SambaStudio](https://python.langchain.com/docs/integrations/chat/sambastudio/), look they have some tiny conflicts with your comments. Is there any up-to-date reference template to look up?

File: docs/docs/integrations/chat/naver.ipynb (hunk 8 above)
Comment (hyper-clova) [reply docs/docs/integrations/chat/naver.ipynb:29]: fixed all links rather than deleting it.

File: docs/docs/integrations/chat/naver.ipynb (hunk 7 above)
Comment (hyper-clova) [reply docs/docs/integrations/chat/naver.ipynb]: now fixed!

**Comment by hyper-clova** [comment]
Hi @efriis , as mentioned above in resolved conversations, we went through over our PR and updated docs and codes. Are there any additional concerns to be merged? Appreciate for your efforts on following up with this PR!

**Comment by hyper-clova** [comment]
Hi @vbarda , just found out that assignee to this pr has changed to you. Looking forward to cooperate!

It seems there was the only lint fail due to the updates in poetry.lock file for the recent test and now fixed. 
Think we are almost there, but if there is any task of change needed, don't hesitate to reach out and let us know! :)

File: libs/community/tests/unit_tests/chat_models/test_naver.py [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+from unittest.mock import patch
+
+import pytest
+from langchain_core.callbacks import BaseCallbackHandler
+from langchain_core.messages import (
+    AIMessage,
+    HumanMessage,
+    SystemMessage,
+)
+from langchain_core.pydantic_v1 import SecretStr
```
Comment (vbarda) [inline libs/community/tests/unit_tests/chat_models/test_naver.py:15]: any reason we're using pydantic v1 here?

File: docs/docs/integrations/chat/naver.ipynb [hunk 10]
```diff
@@ -0,0 +1,444 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d1ee55bc-ffc8-4cfa-801c-993953a08cfd",
+   "metadata": {},
+   "source": [
+    "## Additional functionalities\n",
+    "\n",
+    "### Fine-tuning\n",
```
Comment (vbarda) [inline docs/docs/integrations/chat/naver.ipynb:301]: could we change this to "Using finetuned models"

File: docs/docs/integrations/providers/naver.mdx [hunk 11]
```diff
@@ -0,0 +1,38 @@
+# NAVER
+
+All functionality related to `Naver` including HyperCLOVA X models, especially via `Naver Cloud` [CLOVA Studio](https://clovastudio.ncloud.com/).
```
Comment (vbarda) [inline docs/docs/integrations/providers/naver.mdx:3]:
```suggestion
All functionality related to `Naver`, including HyperCLOVA X models, particularly those accessible through `Naver Cloud` [CLOVA Studio](https://clovastudio.ncloud.com/).
```

File: libs/community/tests/integration_tests/chat_models/test_naver.py [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+
+from langchain_community.chat_models import ChatClovaX
+
+
+def test_stream() -> None:
+    """Test streaming tokens from ChatClovaX."""
+    llm = ChatClovaX()
+
+    for token in llm.stream("I'm Clova"):
+        assert isinstance(token.content, str)
```
Comment (vbarda) [inline libs/community/tests/integration_tests/chat_models/test_naver.py:11]: i would recommend adding more checks here, for example asserting that the streamed chunks are `AIMessageChunk` objects, etc

Comment (vbarda) [reply libs/community/tests/integration_tests/chat_models/test_naver.py:11]: (similar for other tests)

**Comment by vbarda** [comment]
@hyper-clova hey folks -- the change looks good overall! i added a couple of small comments / suggestions. i think we're very close to merging this!

File: libs/community/tests/integration_tests/chat_models/test_naver.py [hunk 13]
```diff
@@ -8,6 +10,7 @@ def test_stream() -> None:
     llm = ChatClovaX()
 
     for token in llm.stream("I'm Clova"):
+        assert isinstance(token, BaseMessageChunk)
```
[the remark this answers is not in the document]
Comment (vbarda) [reply libs/community/tests/integration_tests/chat_models/test_naver.py:13]: this is not blocking, but would probably be nice to update!

**Review by vbarda — APPROVED** [review]

**Pull request merged**